File: check_dates.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

// Directory containing the Excel files
const directory = '창고별재고-january';

// Get all Excel files
const files = fs.readdirSync(directory)
    .filter(file => file.endsWith('.xlsx'))
    .sort();

console.log(`Found ${files.length} Excel files\n`);

// Dictionary to store filename -> date mapping
const fileDates = {};

// Read first row of each file
files.forEach(filename => {
    const filepath = path.join(directory, filename);
    try {
        // Read the workbook
        const workbook = XLSX.readFile(filepath);
        const sheetName = workbook.SheetNames[0];
        const sheet = workbook.Sheets[sheetName];

        // Get the range of the sheet
        const range = XLSX.utils.decode_range(sheet['!ref']);

        // Read first row
        let dateFound = null;
        const firstRowValues = [];

        // Check all cells in the first row
        for (let col = range.s.c; col <= range.e.c; col++) {
            const cellAddress = XLSX.utils.encode_cell({ r: 0, c: col });
            const cell = sheet[cellAddress];

            if (cell) {
                firstRowValues.push(cell.v);

                // Check if it's a date
                if (cell.t === 'd') {
                    dateFound = cell.v;
                    break;
                } else if (cell.t === 'n' && cell.w) {
                    // Try to parse as date (Excel stores dates as numbers)
                    const excelDate = XLSX.SSF.parse_date_code(cell.v);
                    if (excelDate) {
                        dateFound = new Date(excelDate.y, excelDate.m - 1, excelDate.d);
                        break;
                    }
                } else if (cell.t === 's') {
                    // Try to parse string as date
                    const str = cell.v;
                    const dateFormats = [
                        /(\d{4})-(\d{2})-(\d{2})/,
                        /(\d{4})\/(\d{2})\/(\d{2})/,
                        /(\d{4})\.(\d{2})\.(\d{2})/,
                        /(\d{2})\/(\d{2})\/(\d{4})/,
                    ];

                    for (const regex of dateFormats) {
                        const match = str.match(regex);
                        if (match) {
                            const parsed = new Date(str);
                            if (!isNaN(parsed.getTime())) {
                                dateFound = parsed;
                                break;
                            }
                        }
                    }
                    if (dateFound) break;
                }
            }
        }

        if (dateFound) {
            fileDates[filename] = dateFound;
            console.log(`${filename}: ${dateFound.toISOString().split('T')[0]}`);
        } else {
            console.log(`${filename}: NO DATE FOUND - First row: ${firstRowValues.join(', ')}`);
        }

    } catch (error) {
        console.log(`${filename}: ERROR - ${error.message}`);
    }
});

console.log('\n' + '='.repeat(80));
console.log('ANALYSIS:');
console.log('='.repeat(80));

// Check for missing dates in January
if (Object.keys(fileDates).length > 0) {
    const dates = Object.values(fileDates).sort((a, b) => a - b);
    console.log(`\nDate range: ${dates[0].toISOString().split('T')[0]} to ${dates[dates.length - 1].toISOString().split('T')[0]}`);

    // Determine which year and month to check
    const year = dates[0].getFullYear();
    const month = dates[0].getMonth();

    // Generate all dates in the month
    const daysInMonth = new Date(year, month + 1, 0).getDate();
    const allDatesInMonth = [];

    for (let day = 1; day <= daysInMonth; day++) {
        allDatesInMonth.push(new Date(year, month, day));
    }

    // Check which dates are missing
    const foundDateStrings = new Set(
        Object.values(fileDates).map(d => d.toISOString().split('T')[0])
    );

    const missingDates = allDatesInMonth.filter(
        d => !foundDateStrings.has(d.toISOString().split('T')[0])
    );

    if (missingDates.length > 0) {
        console.log(`\nMISSING DATES (${missingDates.length}):`);
        missingDates.forEach(date => {
            console.log(`  - ${date.toISOString().split('T')[0]}`);
        });
    } else {
        const monthName = dates[0].toLocaleString('default', { month: 'long', year: 'numeric' });
        console.log(`\nAll ${allDatesInMonth.length} dates in ${monthName} are present!`);
    }
}

console.log('\n' + '='.repeat(80));
console.log('FILE -> DATE MAPPING:');
console.log('='.repeat(80));

const sortedEntries = Object.entries(fileDates).sort((a, b) => a[1] - b[1]);
sortedEntries.forEach(([filename, date]) => {
    console.log(`${filename} -> ${date.toISOString().split('T')[0]}`);
});

File: check_dates.py
#!/usr/bin/env python3
import os
from openpyxl import load_workbook
from datetime import datetime, timedelta
import glob

# Directory containing the Excel files
directory = "창고별재고-january"

# Get all Excel files
excel_files = sorted(glob.glob(os.path.join(directory, "*.xlsx")))

print(f"Found {len(excel_files)} Excel files\n")

# Dictionary to store filename -> date mapping
file_dates = {}

# Read first row of each file
for filepath in excel_files:
    filename = os.path.basename(filepath)
    try:
        # Load workbook and get first sheet
        wb = load_workbook(filepath, read_only=True, data_only=True)
        ws = wb.active

        # Get first row
        first_row = next(ws.iter_rows(min_row=1, max_row=1, values_only=True))
        date_found = None

        for value in first_row:
            if value is not None:
                # Try to parse as date
                if isinstance(value, datetime):
                    date_found = value
                    break
                elif isinstance(value, str):
                    # Try to parse string as date
                    for fmt in ["%Y-%m-%d", "%Y/%m/%d", "%m/%d/%Y", "%d/%m/%Y", "%Y%m%d", "%Y.%m.%d"]:
                        try:
                            date_found = datetime.strptime(value, fmt)
                            break
                        except:
                            continue
                    if date_found:
                        break

        wb.close()

        if date_found:
            file_dates[filename] = date_found
            print(f"{filename}: {date_found.strftime('%Y-%m-%d')}")
        else:
            print(f"{filename}: NO DATE FOUND - First row: {first_row}")

    except Exception as e:
        print(f"{filename}: ERROR - {str(e)}")

print("\n" + "="*80)
print("ANALYSIS:")
print("="*80)

# Check for missing dates in January 2025
if file_dates:
    dates = sorted(file_dates.values())
    print(f"\nDate range: {dates[0].strftime('%Y-%m-%d')} to {dates[-1].strftime('%Y-%m-%d')}")

    # Determine which year and month to check
    year = dates[0].year
    month = dates[0].month

    # Generate all dates in January
    start_date = datetime(year, month, 1)
    if month == 12:
        end_date = datetime(year + 1, 1, 1)
    else:
        end_date = datetime(year, month + 1, 1)

    all_january_dates = []
    current = start_date
    while current < end_date:
        all_january_dates.append(current)
        current += timedelta(days=1)

    # Check which dates are missing
    found_dates = set(d.date() for d in file_dates.values())
    missing_dates = [d for d in all_january_dates if d.date() not in found_dates]

    if missing_dates:
        print(f"\nMISSING DATES ({len(missing_dates)}):")
        for date in missing_dates:
            print(f"  - {date.strftime('%Y-%m-%d')}")
    else:
        print(f"\nAll {len(all_january_dates)} dates in {start_date.strftime('%B %Y')} are present!")

print("\n" + "="*80)
print("FILE -> DATE MAPPING:")
print("="*80)
for filename, date in sorted(file_dates.items(), key=lambda x: x[1]):
    print(f"{filename} -> {date.strftime('%Y-%m-%d')}")
